File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.js
function solution(message, spoiler_ranges) {
    const notImportants = getNotImportants(message, spoiler_ranges);
    const importants = new Set();
    
    for (const range of spoiler_ranges) {
        const word = getCompleteWord(range, message);
        const splitWords = word.split(/\s+/);
        for (const w of splitWords) {
            const tw = w.trim();
            if (tw && !notImportants.has(tw)) {
                importants.add(tw);
            }
        }
    }
    
    return importants.size;
}

function getCompleteWord(range, origin) {
    let start = range[0];
    let end = range[1];
    const limit = origin.length;

    if (start < 0 || end >= limit || start > end) return "";

    while (start > 0 && origin[start - 1] !== ' ') {
        start--;
    }
    while (end < limit - 1 && origin[end + 1] !== ' ') {
        end++;
    }

    return origin.substring(start, end + 1).trim();
}

function getNotImportants(origin, spoiler_ranges) {
    const notImportants = new Set();
    const isSpoiled = new Array(origin.length).fill(false);

    for (const range of spoiler_ranges) {
        for (let i = range[0]; i <= range[1]; i++) {
            if (i >= 0 && i < origin.length) {
                isSpoiled[i] = true;
            }
        }
    }

    let i = 0;
    while (i < origin.length) {
        if (origin[i] === ' ') {
            i++;
            continue;
        }

        const start = i;
        while (i < origin.length && origin[i] !== ' ') {
            i++;
        }
        
        const end = i - 1;

        let hasSpoiler = false;
        for (let j = start; j <= end; j++) {
            if (isSpoiled[j]) {
                hasSpoiler = true;
                break;
            }
        }

        if (!hasSpoiler) {
            notImportants.add(origin.substring(start, end + 1));
        }
    }

    return notImportants;
}

File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.java
import java.util.*;

class Solution {
    
    public int solution(String message, int[][] spoiler_ranges) {
        Set<String> not_importants = getNotImportants(message, spoiler_ranges);
        Set<String> importants = new HashSet<>();
        
        for (int[] range : spoiler_ranges) {
            String word = getCompleteWord(range, message);
            String[] splitWords = word.split("\\s+");
            for(String w : splitWords){
                String tw = w.trim();
                if (!tw.isEmpty() && !not_importants.contains(tw)) {
                    importants.add(tw);
                }    
            }
        }
        
        return importants.size();
    }

    public String getCompleteWord(int[] range, String origin) {
        int start = range[0];
        int end = range[1];
        int limit = origin.length();

        if (start < 0 || end >= limit || start > end) return "";

        while (start > 0 && origin.charAt(start - 1) != ' ') {
            start--;
        }
        while (end < limit - 1 && origin.charAt(end + 1) != ' ') {
            end++;
        }

        return origin.substring(start, end + 1).trim();
    }

    public Set<String> getNotImportants(String origin, int[][] spoiler_ranges) {
        Set<String> not_importants = new HashSet<>();
        boolean[] isSpoiled = new boolean[origin.length()];

        for (int[] range : spoiler_ranges) {
            for (int i = range[0]; i <= range[1]; i++) {
                if (i >= 0 && i < origin.length()) {
                    isSpoiled[i] = true;
                }
            }
        }

        int i = 0;
        while (i < origin.length()) {
            if (origin.charAt(i) == ' ') {
                i++;
                continue;
            }

            int start = i;
            while (i < origin.length() && origin.charAt(i) != ' ') {
                i++;
            }
            
            int end = i - 1;

            boolean hasSpoiler = false;
            for (int j = start; j <= end; j++) {
                if (isSpoiled[j]) {
                    hasSpoiler = true;
                    break;
                }
            }

            if (!hasSpoiler) {
                not_importants.add(origin.substring(start, end + 1));
            }
        }

        return not_importants;
    }
}